modules/perSystem.nix

```diff
@@ -35,17 +39,41 @@ in
         };
       });
       apply = modules: system:
-        (lib.evalModules {
-          inherit modules;
-          prefix = [ "perSystem" system ];
-          specialArgs = {
-            inherit system;
-          };
-        }).config;
+        let
+          inherit (lib.evalModules {
+            inherit modules;
+            prefix = [ "perSystem" system ];
+            specialArgs = {
+              inherit system;
+            };
+          }) options;
+        # Almost like `config`, except undefined options are filtered away.
+        # This definition is incomplete, but is good enough for our purposes.
+        # Limitations:
+        # - Doesn't handle undefined options in submodules.
+        # - Doesn't handle undefined options that have a prefix, e.g. `_module.args`.
+        # TODO: Upstream a good version of this to Nixpkgs, probably as an extra result
+        # from `evalModules`, e.g. `result.configFiltered`.
+        #
+        # The reason we do this is that we don't want to evaluate the definitions,
+        # and only check whether they're defined. Hence we can't check for nullness,
+        # because that will evaluate a lot of stuff unnecessarily and cause errors
+        # with IFD. A more proper fix would perhaps be fixing the Nix evaluator to
+        # short circuit more, specifically, in an expression like `null != { ${x} = ...; }`,
```

Comment:
This requires a `Value` to have an intermediate state between thunk and attrset, and it will cause make the `!=` operator asymmetric. Neither is terrible in itself, but `nix` has to strike a balance. If you want to make this happens, please contribute to the test suite or other quality control measures such as combining boehmgc and c++ sanitizers.